token/program-2022/src/instruction.rs

```diff
@@ -409,6 +427,126 @@ pub enum TokenInstruction {
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   * Single owner/delegate
+    ///   0. `[writable]` The source account. Must include the `AccountTransferFeeMixin`.
+    ///   1. `[]` The token mint. Must include the `AccountMintFeeMixin`.
+    ///   2. `[writable]` The destination account. Must include the `AccountTransferFeeMixin`.
+    ///   3. `[signer]` The source account's owner/delegate.
+    ///
+    ///   * Multisignature owner/delegate
+    ///   0. `[writable]` The source account.
+    ///   1. `[]` The token mint.
+    ///   2. `[writable]` The destination account.
+    ///   3. `[]` The source account's multisignature owner/delegate.
+    ///   4. ..4+M `[signer]` M signer accounts.
+    TransferCheckedWithFee {
+        /// The amount of tokens to transfer.
+        amount: u64,
+        /// Expected number of base 10 digits to the right of the decimal place.
+        decimals: u8,
+        /// Expected fee assessed on this transfer, calculated off-chain based on
+        /// the transfer_fee_basis_points and maximum_fee of the mint.
+        fee: u64,
+    },
+    /// Transfer all withheld tokens to a fee account. Signed by the mint's
+    /// fee withdraw authority.
+    ///
+    /// Succeeds for frozen accounts.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   * Single owner/delegate
+    ///   0. `[writable]` The token mint. Must include the `MintTransferFeeMixin`.
+    ///   1. `[writable]` The fee receiver account. Must include the `AccountTransferFeeMixin`.
+    ///      associated with the provided mint.
+    ///   2. `[signer]` The mint's `fee_withdraw_authority`
+    ///   3. ..3+N `[writable]` (Optional) The source accounts to harvest from.
+    ///      Must include the `AccountTransferFeeMixin`.
+    ///
+    ///   * Multisignature owner/delegate
+    ///   0. `[writable]` The token mint.
+    ///   1. `[writable]` The destination account.
+    ///   2. `[]` The source or destination account's multisignature owner/delegate.
+    ///   3. ..3+M `[signer]` M signer accounts.
+    ///   3+M+1. ..3+M+N`[writable]` (Optional) The source accounts to harvest from.
+    HarvestFee,
+    /// Close an account by transferring all its SOL to the destination account.
+    ///
+    /// Accounts with withheld transfer fees are transferred into the mint on
+    /// closing.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   * Single owner
+    ///   0. `[writable]` The account to close.
+    ///   1. `[writable]` The destination account.
+    ///   2. `[writable]` The mint.
+    ///   3. `[signer]` The account's owner.
+    ///
+    ///   * Multisignature owner
+    ///   0. `[writable]` The account to close.
+    ///   1. `[writable]` The destination account.
+    ///   2. `[writable]` The mint.
+    ///   3. `[]` The account's multisignature owner.
+    ///   4. ..4+M `[signer]` M signer accounts.
+    CloseAccountWithTransferFee,
```

Comment:
`CloseAccountWithTransferFee` isn't sitting well with me.  What if instead there was a permissionless version of `HarvestFee` that only accepted the writable Mint and writable list of source Accounts to harvest from.

To close an account the client then just needs to add a `HarvestFee` instruction before the standard `CloseAccount` instruction. Some future extension may impose additional/different requirements on account close and then we end up with a zoo of `CloseAccountWith...` variants.

Reply:
Haha this was my initial idea too with `TransferWithheld`!  I am worried about a totally permissionless version of `HarvestFee` -- giving anyone the possibility to move other people's tokens doesn't make me feel great.  But as long as the only possible destination is the Mint, I think it should limit the attack possibilities.

Happy to do this, we just need a good name... `TransferWithheldFeeToMint`? `HarvestFeeToMint`? `ConsolidateFees`?

Reply:
I went with `HarvestFeeToMint` and `HarvestFeeFromMint`